plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,455 @@
+            if (closed) {
+                // Ignore writes after writer is closed
+                return immediateVoidFuture();
+            }
+
+            // Skip multipart upload if there would only be one part
+            if (slice.length() < blockSize && multiPartUploadFutures.isEmpty()) {
+                directUploadFuture = asVoid(toListenableFuture(blockBlobAsyncClient.upload(Flux.just(slice.toByteBuffer()), slice.length()).toFuture()));
+                return directUploadFuture;
+            }
+
+            String blockId = Base64.getEncoder().encodeToString(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+            ListenableFuture<Void> uploadFuture = toListenableFuture(blockBlobAsyncClient.stageBlock(blockId, Flux.just(slice.toByteBuffer()), slice.length()).toFuture());
+            multiPartUploadFutures.add(uploadFuture);
+            blockIds.add(blockId);
+            return uploadFuture;
+        }
+
+        @Override
+        public ListenableFuture<Void> finish()
+        {
+            if (closed) {
+                return immediateVoidFuture();
+            }
+
+            if (multiPartUploadFutures.isEmpty()) {
+                return requireNonNullElseGet(directUploadFuture, Futures::immediateVoidFuture);
+            }
+
+            ListenableFuture<Void> finishFuture = Futures.transformAsync(
+                    Futures.allAsList(multiPartUploadFutures),
+                    ignored -> asVoid(toListenableFuture(blockBlobAsyncClient.commitBlockList(blockIds).toFuture())),
+                    directExecutor());
+            Futures.addCallback(finishFuture, new FutureCallback<>() {
+                @Override
+                public void onSuccess(Void result)
+                {
+                    closed = true;
+                }
+
+                @Override
+                public void onFailure(Throwable ignored)
+                {
+                    // Rely on caller to abort in case of exceptions during finish
+                }
+            }, directExecutor());
+            return finishFuture;
+        }
+
+        @Override
+        public ListenableFuture<Void> abort()
+        {
+            if (closed) {
+                return immediateVoidFuture();
+            }
+            closed = true;
+
+            if (multiPartUploadFutures.isEmpty()) {
+                if (directUploadFuture != null) {
+                    directUploadFuture.cancel(true);
+                }
+                return immediateVoidFuture();
+            }
+
+            verify(directUploadFuture == null);
+            multiPartUploadFutures.forEach(future -> future.cancel(true));
+
+            // No explicit way to delete staged blocks; uncommitted blocks are automatically deleted after 7 days
+            return immediateVoidFuture();
+        }
+
+        @Override
+        public long getRetainedSize()
+        {
+            return INSTANCE_SIZE;
```

Comment:
nit: there are two lists also - but I guess it is fine to skip those as size should be negligible.

Reply:
Good catch. Added accounting for `blockIds`, but not sure how to get estimated size of a `ListenableFuture<Void>`